File: 652.寻找重复的子树.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import javax.swing.tree.TreeNode;

/*
 * @lc app=leetcode.cn id=652 lang=java
 *
 * [652] 寻找重复的子树
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 * int val;
 * TreeNode left;
 * TreeNode right;
 * TreeNode() {}
 * TreeNode(int val) { this.val = val; }
 * TreeNode(int val, TreeNode left, TreeNode right) {
 * this.val = val;
 * this.left = left;
 * this.right = right;
 * }
 * }
 */
class Solution {
    private Map<String, Integer> mp = new HashMap<>();
    private List<TreeNode> ans = new ArrayList<>();

    public List<TreeNode> findDuplicateSubtrees(TreeNode root) {
        serialize(root);
        return ans;
    }

    String serialize(TreeNode root) {
        if (root == null) {
            return "";
        }

        StringBuilder sb = new StringBuilder();

        String left = serialize(root.left);
        String right = serialize(root.right);
        sb.append(left).append("#").append(right).append("#").append(root.val);
        String s = sb.toString();

        mp.put(s, mp.getOrDefault(s, 0) + 1);

        if (mp.get(s) == 2) {
            ans.add(root);
        }

        return s;
    }
}
// @lc code=end

File: 652.寻找重复的子树.cpp
/*
 * @lc app=leetcode.cn id=652 lang=cpp
 *
 * [652] 寻找重复的子树
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
private:
    unordered_map<string, int> mp;
    vector<TreeNode *> ans;

public:
    vector<TreeNode *> findDuplicateSubtrees(TreeNode *root) {
        serialize(root);
        return ans;
    }

    string serialize(TreeNode *root) {
        if (root == nullptr) return "";

        string left = serialize(root->left);
        string right = serialize(root->right);

        string s = left + "#" + right + "#" + to_string(root->val);

        if (mp[s] == 1) {
            ans.emplace_back(root);
        }
        mp[s]++;
        return s;
    }
};
// @lc code=end
